modules/create_timepoint/php/ajaxtimepoint.class.inc

```diff
@@ -0,0 +1,309 @@
+    {
+        // Verify cand ID.
+        $values['identifier'] = isset($values['identifier']) ?
+            htmlentities(
+                $values['identifier'],
+                ENT_QUOTES,
+                'UTF-8'
+            ) : '';
+        return $values;
+    }
+
+    /**
+     * Processes the values and saves to database
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     * @throws \DatabaseException
+     * @throws \DeprecatedException
+     * @throws \LorisException
+     */
+    function processRequest(array $values) : array
+    {
+        if (isset($values['command']) && $values['command'] == 'initialize') {
+            $response = $this->initializeSetup($values);
+        } else if (isset($values['command']) && $values['command'] == 'create') {
+            $errors = $this->validate($values);
+            if (!$errors) {
+                try {
+                    \TimePoint::createNew(
+                        $values['identifier'],
+                        $values['subproject'],
+                        $values['visit'],
+                        isset($values['psc']) ?? null
+                    );
+                } catch (\Exception $exception) {
+                    $exception = '';
+                }
+                $response['status'] = 'success';
+            } else {
+                $response['status'] = 'error';
+                $response['errors'] = $errors;
+            }
+        } else {
+            $response['status'] = 'error';
+        }
+        return $response;
+    }
+
+    /**
+     * Initialize setup, the extra values for the
+     * create timepoint form. (psc & errors)
+     *
+     * @param array $values the user data received.
+     *
+     * @return array
+     * @throws \DatabaseException
+     * @throws \DeprecatedException
+     * @throws \LorisException
+     */
+    function initializeSetup(array $values) : array
+    {
+        // Setup variables
+        $errors         = array(); // form errors.
+        $config         =& \NDB_Config::singleton();
+        $candidate      =& \Candidate::singleton($values['identifier']);
+        $allSubprojects = \Utility::getSubprojectList();
+
+        // Frontend needs for select element.
+        $values['subproject'] = $allSubprojects;
+
+        // All subprojects from config file (error).
+        if (empty($allSubprojects)) {
+            $errors['subprojectID'] = 'No subprojects have been defined 
+            for this study. If you are an administrator, please use the 
+            Configuration module to add new subprojects.';
+        }
+        // List of valid subprojects for a given project
+        if ($config->getSetting('useProjects') === 'true') {
+            // If projects enabled, loop through and add valid subprojects only
+            $subprojList = '';
+            try {
+                $subprojList = $candidate->getValidSubprojects();
+            } catch (\DatabaseException $e) {
```

Comment:
Can you explain why you're silencing the exception here?

Reply:
@johnsaigle the DatabaseException if thrown would pollute the return JSON object and break the frontend (I believe). The line after checks if subprojList is empty and handles the error message that is displayed to the user.

Reply:
It should go inside the `catch` then, I think.

Reply:
Hey @johnsaigle, I looked at the method getValidSubprojects() and it no longer throws the DatabaseException. So I removed the catch in the code and the @throw tag from the method comment tag as well.